Renesas: Smartbond: Add Display Driver Support

ioannis-karachalios opened the pull request.

ioannis-karachalios committed:
dts: renesas: smartbond: Add support for the display driver class.

Define a single node that reflects the LCDC IP. Instead of defining
the same IP block twice with different compatibles (mipi dbi, display)
we define a single node for the default display interface and
other interfaces like the MIPI DBI should override the compatible entry
with the appropriate one within its DTS overlay file.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

danieldegrasse commented on drivers/display/display_smartbond.c:
```diff
@@ -0,0 +1,634 @@
+	(((_val) << LCDC_LCDC_LAYER0_OFFSETX_REG_ ## _field ## _Pos) &	\
+	LCDC_LCDC_LAYER0_OFFSETX_REG_ ## _field ## _Msk)
+
+#if (DT_INST_PROP(0, pixel_format) == PANEL_PIXEL_FORMAT_RGB_565)
```
Instead of using this approach, we should use the `DISPLAY_BITS_PER_PIXEL` macro to determine the pixel size

danieldegrasse commented on drivers/display/display_smartbond.c:
```diff
@@ -0,0 +1,634 @@
+
+	k_sem_take(&data->device_sem, K_FOREVER);
+
+	/* XXX: Unsure if this feature is available in LCDC IP */
```
Can this comment be updated? I'd prefer if this at least highlighted what the bit was doing, and that is might not be present on all IP revisions (assuming I'm understanding the current comment right)

ioannis-karachalios replied:
Sure. This bit will force LCD controller's output to blank that is, the controller will keep operating without outputting any pixel data. Think is, that functionality is not present on all IP revisions and this is why I added some comments.

danieldegrasse commented on drivers/display/display_smartbond.c:
```diff
@@ -0,0 +1,634 @@
+		}
+	}
+
+	/* XXX: Unsure if this feature is available in LCDC IP */
```
Same comment as above, please expand this comment block slightly

danieldegrasse commented on drivers/display/display_smartbond.c:
```diff
@@ -0,0 +1,634 @@
+	return -ENOTSUP;
+}
+
+static int display_smartbond_set_pixel_format(const struct device *dev,
```
It seems this function will error out unless the app is setting the same pixel format it already set at initialization time, which does not seem like something an application would do. Is there some testcase that will fail if we drop this implementation?

ioannis-karachalios replied:
Dropping this implementation would result in failing the character frame buffer (CFB) tests. However, this is not supported by our IP and so we could safely drop this implementation.

danieldegrasse commented on drivers/display/display_smartbond.c:
```diff
@@ -0,0 +1,634 @@
+	}
+
+	/* Make sure continus mode is enabled */
+	LCDC->LCDC_MODE_REG |= LCDC_LCDC_MODE_REG_LCDC_MODE_EN_Msk;
```
Note- I'm not familiar with this IP, just curious- do we need to set this bit here? IE is there some condition that would cause the IP block to disable continuous mode while executing, and require us to force it back on? It seems like once we turn continuous mode on, it would stay on.

Also- spelling of comment above this line needs to be fixed

ioannis-karachalios replied:
As I mentioned below, this is a leftover from previous implementation. I will move this to the initialization routine.

danieldegrasse commented on drivers/display/display_smartbond.c:
```diff
@@ -0,0 +1,634 @@
+	 * If continuous mode is enabled, wait for the current frame to finish.
+	 * Do not disable continuous mode as this will have visual artifacts.
+	 */
+	if (LCDC->LCDC_MODE_REG & LCDC_LCDC_MODE_REG_LCDC_MODE_EN_Msk) {
```
Do we need to check for continuous mode here? Again similar to my comment below- I'm not familiar with the IP, just trying to understand how the continous mode might get disabled at runtime assuming this driver is the only one using the LCDC block

ioannis-karachalios replied:
Correct, there is no need to check for continuous mode here. This is a leftover from previous implementation where continuous mode could be disabled under certain circumstances.

danieldegrasse commented on drivers/display/display_smartbond.c:
I think we might want to include the name of the IP block in this file, in case renesas releases additional display IP blocks. Personally, I would suggest `display_renesas_lcdc.c`- the reasoning being that this will enable the driver to be reused on another family of SOCs if the IP is present there, and it will keep the driver name distinct from other display IP from renesas

danieldegrasse commented on drivers/display/Kconfig.smartbond:
```diff
@@ -0,0 +1,18 @@
+	help
+	  Enable Smartbond display controller.
+
+config DISPLAY_BUFFER_PSRAM
```
This Kconfig should depend on `DISPLAY_SMARTBOND`, and should be namespaced- IE `DISPLAY_SMARTBOND_BUFFER_PSRAM`

danieldegrasse commented on drivers/display/Kconfig.smartbond:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2023 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config DISPLAY_SMARTBOND
```
Same comment as below regarding naming of this file/Kconfig symbol- I think including the IP block name here would make sense, to provide future flexibility

danieldegrasse commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -16,6 +16,10 @@
 		zephyr,flash-controller = &flash_controller;
 	};
 
+	lvgl_pointer {
```
I think this declaration belongs at the board level, not the SOC level

danieldegrasse commented on boards/arm/da1469x_dk_pro/Kconfig.defconfig:
Please split the SOC DTS changes and board DTS changes into separate commits

danieldegrasse commented on samples/modules/lvgl/demos/boards/da1469x_dk_pro.conf:
If there are no Kconfigs we need to set in this overlay file, we should remove it to minimize the number of overlay files in samples

danieldegrasse commented on samples/modules/lvgl/demos/boards/da1469x_dk_pro.overlay:
```diff
@@ -0,0 +1,59 @@
+    pixel-format = <PANEL_PIXEL_FORMAT_RGB_565>;
+
+    /* Display settings for ST7282T2 conttoller */
+    display-timings {
```
Might be worth adding a comment here describing the display model (IE part number) you are using, so others know which module has been tested with this sample

Even better, adding documentation to the board `index.rst` highlighting which display modules are validated with the EVK would probably help quite a bit

ioannis-karachalios replied:
I'd prefer to update board's `index.rst` once the memory controller driver [PR](https://github.com/zephyrproject-rtos/zephyr/pull/68023) is merged (required as most of the display panels resolutions exceed device's system RAM) so I can add all the display part numbers I have tested with.

danieldegrasse commented on tests/drivers/display/display_read_write/boards/da1469x_dk_pro.overlay:
```diff
@@ -0,0 +1,56 @@
+        zephyr,display = &display;
+    };
+
+    // aliases {
```
Need to drop these commented blocks in the overlay

ioannis-karachalios replied:
OK. As I explained before, the reason I've commented out this part is that the MEMC driver has yet to be merged into main.

danieldegrasse commented on tests/drivers/display/display_read_write/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+source "Kconfig.zephyr"
+
+config DISPLAY_READ_WRITE_BUF_SRAM_EXT
```
Why are you introducing this Kconfig, and then setting it to disabled for the only board being added here? Was the Kconfig originally intended to be enabled by this board?

If the Kconfig is no longer needed, I'd prefer that we drop it

ioannis-karachalios replied:
I introduced this Kconfig option so frame buffers can be allocated into external storage mediums, e.g. PSRAM, in case the SoC is running out of system RAM. The reason I have disabled that Kconfig in our board's configuration file is that the memory controller (MEMC) driver implementation has yet to be merged into main ([here](https://github.com/zephyrproject-rtos/zephyr/pull/68023) is the corresponding PR). Should I move this commit to the mentioned MEMC PR?

danieldegrasse replied:
This still needs to be dropped from this PR- it should be added in a follow up PR, once we have MEMC support for the Smartbond part

ioannis-karachalios replied:
Dropped

ioannis-karachalios committed:
boards: renesas: smartbond: Add support for the display driver class.

Update board's DTS configurations to support the Renesas LCD controller.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

danieldegrasse commented on boards/renesas/da1469x_dk_pro/Kconfig.defconfig:
```diff
@@ -5,8 +5,37 @@
+if LVGL
+
+# LCDC imposes display buffer be word aligned
+config LV_Z_VDB_ALIGN
```
This is the default value, and can be dropped.

danieldegrasse commented on drivers/display/display_renesas_lcdc.c:
```diff
@@ -0,0 +1,595 @@
+	return ret;
+}
+
+static int display_smartbond_set_orientation(const struct device *dev,
```
Is this implementation needed? The display API no longer requires all functions be implemented, it will return `-ENOSYS` if one isn't present.

ioannis-karachalios committed:
drivers: display: smartbond: Add support for the display driver class.

Exhibit Renesas LCD controller's driver implementation. The driver
is intended to employ the controller in the continuous mode so
it can drive display panels in the parallel RGB mode.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

danieldegrasse commented on drivers/display/display_renesas_lcdc.c:
```diff
@@ -0,0 +1,595 @@
+	 * directly by LCDC and is allocated statically during device initialization. The color
+	 * format is defined based on the pixel-format property dictated by lcd-controller
+	 * bindings.
+	 */
```
Please be consistent here as far as multi-instance versus single instance code. For example, if you wanted to make this block entirely multi instance, it would look something like this:

```
capabilities->supported_pixel_formats = config->pixel_format
capabilities->current_orientation = DISPLAY_ORIENTATION_NORMAL;
capabilities->current_pixel_format = config->pixel_format
capabilities->x_resolution = config->x_res;
capabilities->y_resolution = config->y_res;
```

Alternatively, an entirely single instance version would look like this:

```
capabilities->supported_pixel_formats = DT_INST_PROP(0, pixel_format);
capabilities->current_orientation = DISPLAY_ORIENTATION_NORMAL;
capabilities->current_pixel_format = DT_INST_PROP(0, pixel_format);
capabilities->x_resolution = DT_INST_PROP(inst, width);
capabilities->y_resolution = DT_INST_PROP(inst, height);
```

Personally, I would *strongly* prefer the driver is implemented as multi instance, since that is more scalable in the future.

danieldegrasse commented on drivers/display/display_renesas_lcdc.c:
```diff
@@ -0,0 +1,595 @@
+				const uint16_t x, const uint16_t y,
+				const struct display_buffer_descriptor *desc,
+				const void *buf)
+{
```
I'm not familiar with this LCD controller, but something that has been implemented for the ELCDIF (which is designed to stream a buffer in memory directly to the display) is to switch to the user provided buffer directly if the buffer is the same size as the output display.

https://github.com/zephyrproject-rtos/zephyr/blob/eaf140e7ff063c2100f8c7dc4402cf6f7ba44506/drivers/display/display_mcux_elcdif.c#L94

I'm not sure if this is something your controller supports, but it might be worth investigating- it improved performance quite a bit using LVGL with the ELCDIF

danieldegrasse replied:
> What is also not clear to me is, if LCDC employs application buffer directly then any rending performed during frame update shouldn't result in visual artifacts?

The trick here would be to queue the application buffer somehow, and not load it into the LCDC hardware until the last frame was done streaming. Again, I am *not* familiar with the hardware here- so this should be taken as an idea, nothing more. No need to implement it if it is not possible or not worth looking into

ioannis-karachalios replied:
I see - such an architectural scheme should not work for us. However, I will spend some time on the underlying implementation :)

danieldegrasse commented on samples/drivers/display/boards/da1469x_dk_pro_lcdc.overlay:
```diff
@@ -0,0 +1,46 @@
+
+	/*
+	 * Panel settings for the NHD-4.3-480272EF-ASXP-CTP
+     * display panel model which integrates the SC7283
```
nit- fix indentation

danieldegrasse replied:
Indentation still looks wrong here

ioannis-karachalios commented:
@danieldegrasse is there any further feedback? Could you remove the `platform: ITE` label added mistakenly after the last rebase?

danieldegrasse commented on samples/modules/lvgl/demos/boards/da1469x_dk_pro_lcdc.overlay:
I think if we are going to add overlays for each display supported with this board, it really might be valuable to consider moving these to the board `dts` folder. There are other display samples in tree (and I suspect some customers will develop using the development kit and this display initially), so putting the overlays in the board's `dts` folder will simplify supporting other display samples or customer applications

ioannis-karachalios replied:
There should be parts that are not common for all sample codes/applications e.g. LVGL and input pointer. Then, one should create another overlay under application's `board` folder to add the extra parts, isn't it? In addition, any overlay should be explicitly invoked during `west build` since multiple overlays are expected per LCD interface type i.e. da1469x_dk_pro_lcdc.overlay, da1469x_dk_pro_mipi_dbi.overlay. This is something we discussed in the MIPI DBI PR since the same LCDC IP is used for different driver classes. Therefore, this might require defining two overlay files in DTC_OVERLAY_FILE isn't it?

danieldegrasse replied:
Sure- but defining the LVGL/input pointer in an application that does not use `CONFIG_LVGL=y` will have no effect. I think you could probably simplify this to two overlays, one for the LCDC IP using the MIPI DBI driver class, and one for the LCDC IP using the display driver class.

ioannis-karachalios replied:
OK - I was a bit of unsure whether or not defining blocks that are not used would be be acceptable.

ioannis-karachalios committed:
boards: renesas: dts: Add overlay files to support LCDC

In order to avoid defining almost the same overlays for different
sample codes and/or applications a common overlay file per
display interface is defined under the boards dts folder.
In doing so, an application/sample code will only have to
define another overlay explicitly under application's board
folder if more blocks are to be enabled. In either case, users
should explicitly invoke the requested overfiles at 'west build'
invokation by using the DTC_OVERLAY_FILE system variable.

Signed-off-by: Ioannis Karachalios <ioannis.karachalios.px@renesas.com>

danieldegrasse commented on dts/arm/renesas/smartbond/da1469x.dtsi:
```diff
@@ -332,6 +332,13 @@
 			interrupts = <9 0>;
 		};
 
+		display: lcdc@30030000 {
```
Instead of defining the same IP block twice with different compatibles, you should override the compatible of the device in the overlay file, something like so:

da1469x.dtsi:
```
lcdc: lcdc@30030000 {
    	compatible = "renesas,smartbond-display";
	reg = <0x30030000 0x18C>;
	interrupts = <32 0>;
	status = "disabled";
};
```
da1469x_dk_pro_mipi_dbi.overlay
```
&lcdc {
    status = "okay";
    compatible = "renesas,smartbond-mipi-dbi";
};
```

ioannis-karachalios replied:
Thanks for the hint. I followed your suggestion but I am now unable to build. In specific, when I try to build for the display compatible via
`west build -p always -b da1469x_dk_pro samples/drivers/display/ -- -DDTC_OVERLAY_FILE="da1469x_dk_pro_lcdc.overlay"` I am getting errors on missing properties of the MIPI DBI compatible. The same goes if I build for the MIPI DBI compatible. Could you please check?
<img width="953" alt="display_build_error" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/bdedc651-e945-4b9e-bf1d-9244d975ddc4">
<img width="956" alt="mipi_dbi_build_error" src="https://github.com/zephyrproject-rtos/zephyr/assets/13284615/a8ed67b0-8216-49f9-b585-6123fcd3a2b5">

ioannis-karachalios replied:
Forget my previous comment; I messed up the compatible properties between the display and MIPI DBI overlays :/

danieldegrasse replied:
No worries! Thanks for getting this changed :)

ioannis-karachalios commented:
@danieldegrasse many thanks for your contribution to this PR.
